### ISSUE TITLE
Possibly bad/incomplete libzmq+libsodium DLL:s (Windows)

### ISSUE THREAD

**Issue by amonamarth** [zeromq/clrzmq4#62]
TL;DR: The libzmq DLL in 4.1.0.21 is incomplete and/or not linking properly with libsodium (libsodium-related functions in libzmq set errno "ENOTSUP"). Suggest fixed libzmq DLL or revert to older version of libzmq DLL.

---

I am using Visual Studio 2015 (using .NET 4.5.2.), with ZeroMQ/clrzmq4 version 4.1.0.21 (installed with NuGet).

A System.InvalidOperationException is thrown when calling ZeroMQ.Z85.CurveKeypair (and other libsodium dependent functionality). Catching the exception and checking the ZeroMQ.lib.zmq.errno(), "ENOTSUP" is returned. This indicates that the libzmq DLL was built without libsodium/curve support. 

A colleague used an earlier version of the ZeroMQ bindings (the exact version escapes me) installed with NuGet. In his case, the call worked fine. Since he was using NET 4.0, we tried to update his ZeroMQ version to 4.1.0.21 to see if the call still worked (it didn't). 

Since I figured one of the DLLs where to blame, I tried replacing them with ones I built myself from jedisct1/libsodium and zeromq/libzmq, but it didn't work. It seems libzmq has changed the way in which the DLL builds. According to my colleague it previously linked statically with libsodium (or at least it was a build option), but now that option is gone. When building the libzmq.dll you are left with the libsodium and libzmq dlls where,the libzmq dll does not load or dynamically link with the functions in the libsodium dll.

I basically solved the problem by tweaking the libzmq build by
1) setting a preprocessor definition that was not set when building libzmq together with libsodium (I think it was ZMQ_HAVE_CURVE). It is not properly set by configure.bat when building for msvc.
2) making sure to build libzmq.dll and link it to libsodium.lib (this is a dynamic import library containing stubs to load the correpsonding functions in libsodium.dll)

With the new dlls, the call to ZeroMQ.Z85.CurveKeypair call worked. I have only tested this with Visual Studio 2015, so I don't know if the build is broken for all Visual Studio versions (both libzmq and libsodium have separate msvc solutions for each version).

I don't know if there is a way to handle the linking issue directly in ZeroMQ, for example with some kind of interop/dll import trickery.

Hopefully this is not a waste of your time :)

**Comment by metadings** [comment]
TL;DR: You can **happily** replace the files, preferrably using **new** files, in folders `i386` and `amd64`, so to say `libzmq.dll` and `libsodium.dll`, also their corresponding `.so` files.

---

Well no... Actually I don't have a Windows machine, so I can't make you a new build of `libzmq` and `libsodium`... (which I thought I did make with `libsodium` and `libpgm`...)

Reverting the update means you get a lot **older** version of `libzmq` which is, let's face it: bullshit.

I've made the update, to just have one and I've used .NET 4.0 and VC2010 to run this application on every computer on this earth.

However, you are using VS2015 (latest ... ;)), you may also try to build the `C` project using VS2012...? I'm not really sure about wich runtime I should support; you may deside so.

Then you `fork` [github.com/zeromq/clrzmq4](https://github.com/zeromq/clrzmq4), so you have `github.com/amonamarth/clrzmq4`...
Then do `git clone https://github.com/amonamarth/clrzmq4`.
Do change the files `libzmq.dll` and `libsodium.dll` in folders `i386` and `amd64`.
Finally, you can `git push origin master` and send me a "Pull Request" to let me import (git pull) your changes.

**Comment by metadings** [comment]
I'm going to tell you: open your Visual Studio, or MonoDevelop, open the project [ZeroMQ/clrzmq4](https://github.com/zeromq/clrzmq4) and build the Project... So to say, you do open [ZeroMQ/libzmq](https://github.com/zeromq/libzmq) (@hintjens), also [jedisct1/libsodium](https://github.com/jedisct1/libsodium) (@jedisct1), and possibly also [steve-o/openpgm](https://github.com/steve-o/openpgm) (@steve-o), and build their projects!

**Comment by amonamarth** [comment]
Maybe I came across as a bit rude, and in that case I'm sorry. I have never contributed to an open source project (I actually created this account just to be able to open an issue here). I wasn't even sure if opening an issue was the right choice, or if it would be better to just make a comment on the commit [Updating v4.1 with curve, epgm (Windows)](https://github.com/zeromq/clrzmq4/commit/71e04a37fb16d6105e6dbb78f3645927725672f9). 

I did "happily" build and replace the dll:s, and they worked for my use case. However, I am not comfortable with forking and committing the dll:s, since I don't have much experience with this and probably made some silly mistake or an error (for example, I didn't even think about the different Visual Studio runtimes). This is the reason I created a new issue; to make sure this is known if someone else has the same problem, and maybe someone with more experience and knowledge would be able to fix it properly.

I have actually been very nervous about even trying to contribute in any way to a project like this, since I fear the response would be hostile.

**Comment by metadings** [comment]
TL;DR: Well no... You are not rude! You didn't say things like "f_ck" or "assh_le"... You said something about `new ZSocket`, `new ZFrame` and the like :-) You are welcome!

---

Well no, I am a one-man-show running some `github.com/zeromq/clrzmq4`, this is about `new ZSocket` and `new ZFrame`... I believe I'm looking like the big-ZeroMQ-ZGuide-things and there; look: it's running on C#! :-)

So no, I am a one-man-show, sometimes doing 

```
// amd64/libzmq.dll, amd64/libsodium.dll
// i386/libzmq.dll, i386/libsodium.dll
// ... changed
git add -A
git commit -m "Updating v4.1 with curve, epgm (Windows)" 
git push zeromq master
```

**Comment by amonamarth** [comment]
Thanks, I needed that :)
Anyways, I found an issue on [libzmq](https://github.com/zeromq/libzmq/issues/1963) regarding the build problem I had with libzmq and libsodium; when it is resolved, I (or someone else I guess :) ) could build the new DLL:s and commit them to a fork of clrzmq4. I'd like to at least see what people have to say about how to resolve the issue in libzmq before I try to build them again (e.g. if I overlooked some important detail when building my current DLL:s) :-)

Also, I would like to show some code, since my previous explanation perhaps was a little unclear.
This code is a short example of what I was trying to do, but it failed.

```
    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                byte[] publicKey;
                byte[] privateKey;
                ZeroMQ.Z85.CurveKeypair(out publicKey, out privateKey);  // <-- throws an System.InvalidOperationException

                Console.WriteLine("public key:\n"+BitConverter.ToString(publicKey));
                Console.WriteLine("private key:\n" + BitConverter.ToString(privateKey));
            }
            catch(Exception e)
            {
                int errnum = ZeroMQ.lib.zmq.errno(); // Returns 127 (ENOTSUP), set by zmq_curve_keypair if libzmq was not built with curve support (i.e. using libsodium)
                String errstr = Marshal.PtrToStringAnsi(ZeroMQ.lib.zmq.strerror(errnum)); // Just says "Not supported"
                Console.WriteLine("errno (" + errnum + "): " + errstr);
                Console.WriteLine(e);
            }     
        }
    }
```

It fails because libzmq.dll was built without support for curve (which is implemented with libsodium I think?). This is because of the build problem in libzmq I mentioned earlier (it seems to affect at least VS2012 and VS2015), not because of a problem in clrzmq4 (maybe I shouldn't have created an issue here?).

Other code that has nothing to do with curve works fine. For example, implementing the [grasslands pattern](http://hintjens.com/blog:49#toc2) (translated to C# of course :) ) works perfectly.

Also, I want to thank you for your work with the binding, because it is really useful :)

**Comment by metadings** [comment]
> I want to thank you for your work with the binding, because it is really useful :)

YEAH :+1:

Well, I don't have a Windows computer to test, nor do I have a test suite, I'm just running `./ZGuideExamples.exe Espresso` to check some features. I have to admit, I didn't check any of the curve features.

You should make a new `libzmq.dll`, with `libsodium.dll` and using `libpgm.lib`, for VS2012 to which also Mozilla has changed recently, and then `git` that to me... 
I'm also not that hard in C/C++, so please don't be too shy to do this :-)

However, if you already installed VS2015, and there may/will be no way to install VS2010 or VS2012; maybe you can use msbuild to get the target right - but I don't want to overrun you with the Setup and Installation mess. So to say, wait for someone to build that thing, and stay tuned!

**Comment by kramerids** [comment]
I'm having the same issue as described above in the original question.  I'm trying to use the Curve security features of libsodium but haven't been able to get it to work.  Will there be an update to the project that includes the sodium enable libzmq in the near future?

If not can someone provide detailed instructions on how to build a portable version of the libraries with sodium enabled?

Looking for any guidance I can get to do this correctly.

**Comment by metadings** [comment]
> Will there be an update to the project that includes the sodium enable libzmq in the near future?

Yes, I'm going to push NuGet/ZeroMQ v4.1.0.22 - with the older version of libzmq v4.1.x, just to have zmq_curve_keypair back.

Please continue this discussion in [Issue #70](https://github.com/zeromq/clrzmq4/issues/70).
